enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,168 @@
+---
+title: remediation-history
+authors:
+  - "@slintes"
+reviewers:
+  - "@beekhof"
+approvers:
+  - "@JoelSpeed"
+  - "@michaelgugino"
+  - "@enxebre"
+creation-date: 2020-12-15
+last-updated: 2020-12-15
+status: implementable
+---
+
+# Remediation history
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Record and show limited remediation history.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an unlimited history of remediations.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was succesfully fenced (deleted / powered off)
+- when the node is healthy again.
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+Example MHC:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  annotations:
+    machine.openshift.io/remediation-strategy: external-baremetal
+  creationTimestamp: "2020-11-24T16:31:22Z"
+spec:
+  maxUnhealthy: 100%
+  nodeStartupTimeout: 60m
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: worker
+  unhealthyConditions:
+  - status: "False"
+    timeout: 20s
+    type: Ready
+  - status: Unknown
+    timeout: 20s
+    type: Ready
+status:
+  conditions:
+  - lastTransitionTime: "2020-11-24T17:17:32Z"
+    status: "True"
+    type: RemediationAllowed
+  currentHealthy: 2
+  expectedMachines: 2
+  remediationHistory:
+  - conditionStatus: Unknown
+    conditionType: Ready
+    detected: "2020-11-24T17:33:43Z"
+    fenced: "2020-11-24T17:34:35Z"
```

Comment:
There is a difference if you look at it from a baremetal perspective yes. But on a cloud provider, the only action we take is to delete the Machine. The Machine is then drained and terminated and we get a _new_ Machine. There's no real sense of fencing in MHC for the cloud case (or even OSP/vSphere). I would also define remediated in this case as MHC has completed its task, which for us is just deletion. We can't say the Node goes healthy again because the node dies.

Reply:
If the point of the `fenced` field is to track when it's in a safe state and that term is only used for metal, maybe `safe` is a better name?

Reply:
In my perspective, `fenced` (not sure if it's the best terminology) is when the cluster can safely reschedule the workloads on other node.
From an admin point of view, `fenced` is the time he should expect that the cluster will start recovery (rescheduling) of the workload.

IMO, it also makes sense for cloud env as well. 
Practically, the workload can be scheduled elsewhere when the node got drained or if it's unreachable - when the node is deleted.
`remediatied` (better than `finished` IMO), should be the time where the compute resource is back and ready for running workload. In cloud environment, it would be the time after a new replacement node has joined the cluster.

> There's no real sense of fencing in MHC for the cloud case (or even OSP/vSphere). 

After machine deletion, the machine controller deletes the compute resource from the provider. This is when fencing practically happens in cloud env.

> I would also define remediated in this case as MHC has completed its task, which for us is just deletion. We can't say the Node goes healthy again because the node dies.

True, but from an admin perspective, I guess that `remediated` is when the compute resource is back, i.e. when the node recovered, or when a new replacement node joined the cluster.

> I guess the point here is that for cloud, fenced and remediated are the same, all we do is delete.

As I mentioned above, I'd suggest that `fenced` would be the time the *node* was deleted (or successfully drained) from the cluster, and `remediated` would be the time a new/fixed node rejoins the cluster.

Maybe `fenced` and `remediated` should be renamed to reflect that.

Reply:
> In cloud environment, it would be the time after a new replacement node has joined the cluster.

We don't track this in MHC and honestly, I don't think we want to. MHC gives up on a machine (in cloud at least) once it's deleted. We have no replacement concept so there would be no way to say a new Node replaced an old broken one.

> After machine deletion, the machine controller deletes the compute resource from the provider. This is when fencing practically happens in cloud env.

So to be correct, should the MHC track the Machine and look for the Deleting phase? I'm concerned this effort may over complicate the MHC. 

> True, but from an admin perspective, I guess that remediated is when the compute resource is back, i.e. when the node recovered, or when a new replacement node joined the cluster.

This isn't something that we track and I'm not sure we can track in a cloud environment, there is no concept of a 1:1 replacement so the only thing that I could see making sense would be when the Machine no longer exists?

Reply:
Yeah, I agree. 
I was mainly taking the customer point of view and less on how things can be implemented.
Not sure if this is worth the effort and complication

Reply:
If the cloud implementation for remediation just deletes a Machine, then how are we going to know when it has been replaced to set the `finished` value? The new Machine will have a different name. The only link to the old Machine will be through the MachineSet, right? But if 2 Nodes based on Machines from the same MachineSet are remediated at the same time how will we know which "replacement" matches which original Machine?

Reply:
The only way I can think `remediated/finished` makes sense for the cloud side would be if we set it to the timestamp we first see the Machine has gone away, ie, reconcile MHC, see status references machine not yet finished, does machine exist, no, set finished

> I was mainly taking the customer point of view and less on how things can be implemented.
> Not sure if this is worth the effort and complication

Makes sense. I would prefer we come up with some way to make this work for all providers and to be consistent, though that is proving tricky, perhaps for a cloud it is a case of `fenced` == `deletionTimestamp` and `remediated` matches what I said at the beginning of this comment? Not married to this idea though

Reply:
Nir explained my initial idea about fenced/remediated/finished better than I did, thanks!
But I understand that it's not as easy than I thought to track everything I wanted to track. I assumed that a new node, which replaces a deleted node, will always have the same name again. Since this is not (always) the case, and following your comments above, I think tracking a `finished/remediated` state is too complicated, and we might just skip it (at least for now, adding fields is always possible). It also wasn't in the requirements, I added it because I thought it will be easy to track. Is this ok for everyone?
About the `fenced/safe` timestamp, quoting Nir: this is "when the cluster can safely reschedule the workloads on other node". In the cloud case this is when the node was deleted.
But I see an issue(?) here as well: while we know that node deletion will happen with the built in remediation strategy, this might not happen with external remediation (e.g. reboot strategy). The fact that external remediation is used is already tracked in the `remediationType` field. Maybe the MHC controller can just leave the fenced/safe field empty in this case, and leave it up to the external remediation controllers to fill it? Thoughts?

Reply:
I'll throw `complete` into the mix and claim that it should be set when either:
- the Node is deleted[1], or 
- the Node returns to a healthy state

[1] IIRC this always happens before the MHC deletes the Machine.

Reply:
Actually (reading further for additional context), what I described applies to "finished".

"detected" should be when the error condition is first observed by MHC
"fenced" (or equivalent) is when the condition persists for longer than the configured timeout and recovery was initiated.  perhaps "triggered" or "recovery initiated" ?
"finished" is when the scheduler can start allocating workloads elsewhere (either because the Node is healthy or has been deleted)

Reply:
Naming things is hard ;) And looking at outdated code snippets does not make it easier...
your "fenced", which you like to be "triggered" or "initiated" is what I call "started"
your "finished" is "remediated" in the latest version.

Reply:
sorry for the confusion... detected, started, and finished looks good to me 👍